hwy/contrib/unroller/unroller-inl.h

```diff
@@ -0,0 +1,496 @@
+    auto maskneg = hn::Not(hn::FirstN(
+        d_in1,
+        static_cast<size_t>(places + static_cast<ptrdiff_t>(UnitLanes()))));
+    if (places < 0) mask = maskneg;
+
+    return hn::MaskedLoad(mask, d_in1, from + idx);
+  }
+
+  inline ptrdiff_t Store(ptrdiff_t const& HWY_RESTRICT idx,
+                         OUT_T* HWY_RESTRICT to,
+                         hn::Vec<OT> const& HWY_RESTRICT x) {
+    return me()->StoreImpl(idx, to, x);
+  }
+
+  inline ptrdiff_t StoreImpl(ptrdiff_t const& HWY_RESTRICT idx,
+                             OUT_T* HWY_RESTRICT to,
+                             hn::Vec<OT> const& HWY_RESTRICT x) {
+    hn::StoreU(x, d_out, to + idx);
+    return d_out.MaxLanes();
+  }
+
+  inline ptrdiff_t MaskStore(ptrdiff_t const& HWY_RESTRICT idx,
+                             OUT_T* HWY_RESTRICT to,
+                             hn::Vec<OT> const& HWY_RESTRICT x,
+                             ptrdiff_t const& HWY_RESTRICT places) {
+    return me()->MaskStoreImpl(idx, to, x, places);
+  }
+
+  inline ptrdiff_t MaskStoreImpl(ptrdiff_t const& HWY_RESTRICT idx,
+                                 OUT_T* HWY_RESTRICT to,
+                                 hn::Vec<OT> const& HWY_RESTRICT x,
+                                 ptrdiff_t const& HWY_RESTRICT places) {
+    const hn::ScalableTag<ptrdiff_t> di;
+    using TI = hn::TFromD<decltype(di)>;
+    auto mask = hn::FirstN(d_out, static_cast<size_t>(places));
+    auto maskneg = hn::Not(hn::FirstN(
+        d_out,
+        static_cast<size_t>(places + static_cast<ptrdiff_t>(UnitLanes()))));
+    if (places < 0) mask = maskneg;
+
+    hn::BlendedStore(x, mask, d_out, to + idx);
+    return std::abs(places);
+  }
+
+  inline ptrdiff_t Reduce(hn::Vec<OT> const& HWY_RESTRICT x,
+                          OUT_T* HWY_RESTRICT to) {
+    return me()->ReduceImpl(x, to);
+  }
+
+  inline ptrdiff_t ReduceImpl(hn::Vec<OT> const& HWY_RESTRICT x, OUT_T* to) {
+    // default does nothing
+    return 0;
+  }
+
+  inline void Reduce(hn::Vec<OT> const& HWY_RESTRICT x0,
+                     hn::Vec<OT> const& HWY_RESTRICT x1,
+                     hn::Vec<OT> const& HWY_RESTRICT x2,
+                     hn::Vec<OT>& HWY_RESTRICT y) {
+    me()->ReduceImpl(x0, x1, x2, y);
+  }
+
+  inline void ReduceImpl(hn::Vec<OT> const& HWY_RESTRICT x0,
+                         hn::Vec<OT> const& HWY_RESTRICT x1,
+                         hn::Vec<OT> const& HWY_RESTRICT x2,
+                         hn::Vec<OT>& HWY_RESTRICT y) {
+    // default does nothing
+  }
+};
+
+template <class FUNC, typename IN_T, typename OUT_T>
+inline void Unroller(FUNC& f, IN_T* HWY_RESTRICT x, OUT_T* HWY_RESTRICT y,
+                     const ptrdiff_t n) {
+  const auto lane_sz = f.UnitLanes();
+
+  auto xx = f.X0Init();
+  auto yy = f.YInit();
+  ptrdiff_t i = 0;
+
+#if HWY_MEM_OPS_MIGHT_FAULT
+  if (n < lane_sz) {
+    // stack is maybe too small for this in RVV?
```

Comment:
We can update this comment: 'this may not fit on the stack for HWY_RVV, but we do not reach this code there'.